Fix dependencies and build on the `romcal-es6`branch

emagnier opened the pull request:
* Update node dependencies.
* The `@std/esm` package has been deprecated in favour of `esm`. Also `esm` have limited the `.mjs` files to basic functionality without support for esm options, to align with the current experimental support in Node. [More details here](https://github.com/standard-things/esm/issues/696). According to that all `.mjs` files in the `romcal` project have been renamed back to `.js`.

The front-end build still need some works, including uglify/minify and configure source-map.

This also might fix #49.

emagnier commented:
@pejulian, I also notice that you didn't transpose the cache for dates, on the new es6 codebase. Is it intentional, or something that you didn't have time to include again?

emagnier commented on src/locales/skSk.js:
```diff
@@ -344,7 +344,7 @@ export default {
     "saintVitusMartyr": "Svätého Víta, mučeníka",  // austria, czechRepublic, germany, switzerland
```
@tukusejssirs do you think we could remove theses comments here in this file? I think it's not really their place. And if needed, you can still find them in the code history. Also I would like to transform theses files to `json`, so we are not tempted to add logic or variables here (and `json` unfortunately don't like comments). And it might be also a step to moving localisation and calendars forward, to make theses ressources more independant from the core codebase, so making them easiest to manage and update.

emagnier replied:
Ok thank you. I'll do it in a separated PR (the `.js` -> `.json` files), and we will discuss about the localisation elsewhere in the issue tracker. I'm also thinking about a separated repo containing calendars and localisations, but I need to dive into it first and do some code/architecture experiments (particularly for the calendars). But it was just a side comment andiIn the meantime we need to merge that PR 😄

tukusejssirs replied:
Okay to everything you said. Still I would note that there is a need to have historical calendars (at the Extra-Ordinary Calendar, which is not a _true_ historical calendar; see #23, #51).

tukusejssirs commented:
Anyway, do you know why this PR is checked by Travis? I cannot find it at [Travis](https://travis-ci.org/pejulian/romcal/builds?utm_medium=notification&utm_source=github_status) too.

emagnier commented on index.html:
```diff
@@ -30,8 +30,11 @@ <h1>Romcal</h1>
       var dates = Romcal.calendarFor({
-        year: 2018
+        year: currentYear,
+          locale: 'fr'
```
yep, in fact this was a test and I forgot to remove it. Good catch!

tukusejssirs replied:
I would suggest againt removing it altogether, but modify it to check the browser locale and use that. `romcal` defaults to `enUs` when that particular language is not available.

emagnier replied:
It's a nice to have that will be added in a separated PR. We also need to check front-end side what locales are available on Romcal. In the meantime I will just remove it.

emagnier replied:
And also because there might need a bit of work backend side: https://github.com/romcal/romcal/issues/26#issuecomment-558401426

pejulian replied:
Just to point out, when I first 

> > @pejulian, I also notice that you didn't transpose the cache feature for dates, on the new es6 codebase. Is it intentional, or something that you didn't have time to include again?
> 
> As @pejulian is currently AWOL, do what you think is the best. 😃

The original intent was to convert the entire code base into Typescript. I must admit however that I only managed a partial conversion of the code due to my limited knowledge at the time. Since then, I have come a long way in Node module development with Typescript and have better ideas of how the code base may be transformed into a type safe structure.

emagnier replied:
Thanks for your feedback @pejulian! Now that I got deeper into the codebase, I was also considering to switch it to Typescript. In the meantime I chose to introduce Flow (in my last PR #158), so we can continue to improve the codebase litle by litle, with the goal to switch then the entire things to Typescript someday.

tukusejssirs commented:
> @pejulian, I also notice that you didn't transpose the cache feature for dates, on the new es6 codebase. Is it intentional, or something that you didn't have time to include again?

As @pejulian is currently AWOL, do what you think is the best. :smiley:

emagnier commented:
### 1 - install / building/ running steps

Generally Node.js projects (like that one) on Github don't specify that you have first to install node.js (which include npm). Because people on Github are mostly developpers. 
But I'm aware that `romcal` isn't a "mainstream" project, and might interest people not all developpers or familiar with JavaScript / Node.js development.

However I dont' think it's a good idea to give all the detailed instructions on how install Node.js, or npm, or Yarn... Because there is various ways in different OS. For exemple I've installed node.js via [`nvm`](https://github.com/nvm-sh/nvm) so I can easily switch between different node.js versions on my mac.

Instead I'll suggest to complete a little bit the "[Usage](https://github.com/romcal/romcal/tree/romcal-es6#usage-)" section in the readme.md, and maybe rename it to Installation instead of Usage (more clear).

The beginning could looks like:

### ⏳ Installation
Install **Romcal** with this Quickstart command :
- Use NPM (with [node installed](https://nodejs.org/en/download/))
```
$ npm install romcal
```
- Use yarn ([install yarn with these docs](https://yarnpkg.com/lang/fr/docs/install/))
```
$ yarn add romcal
```
... and so on.
You don't have to clone romcal, except if you want to contribute to this project (so theses instructions should be elsewhere in the doc).

In the contribution section, we should suggest first to fork `romcal` in our own github account, and then clone (to install locally) this forked repo. So you have your own working space (it's a best practice). But I'm pretty sure that this kind of documentation already exists somewhere on Github, so we don't have to reinvent the wheel.

Also it's no necessary to install `webpack`, `webpack-cli` or `mocha` globally. That should be fixed in this PR.

And I don't recommend to do `npm audit fix`. Since it could update/modify dependencies of dependencies that could break things. It's best to keep dependencies up to date as far as we can, and if there is critical vulnerabilities, we should report them instead, in the issue tracker of this dependencies.

### 2
I don't have this issue one Firefox, Safari or Chrome on my mac. But there is still some fine tuning on the build steps for the web package of `Romcal`. Probably in another PR.

### 3
Here it's just to give a very basic exemple, that we can easily hack directly in the browsers. So I don't think it's a problem if it don't follow good practices in web development here.

### 4
I already created a new version `1.3.1`. The major refactor from @pejulian correspond to `1.3.0` (also not published yet on NPM). 

To stay in the scope of this PR, the following points you are mentioning (as well as the installation documentation) should be handled in separated PRs. 

Travis: I have not looked at it yet. I know it was working and checking automatically all new PR for this project, a few months (years) ago. But I don't know what's going on today. Yes we should took a look into it as well...

emagnier commented:
I've removed the `fr` localisation in the `index.html`, and squashed together all the commits concerning the dependencies and build work. All other suggestions will be addressed in future separated PRs. 
If everything looks good I think we can merge this PR. Thanks!

tukusejssirs commented:
> ### 1 - install / building/ running steps

I have no problem with your suggestions as long as the instructions will be accessible in more detail. I _am_ a `node.js`/`npm` beginner (I have never dealt with any `node.js`/`npm` project before `romcal`). And I really think that `romcal` could have only a _specialised_ community, i.e. especially those who are Roman Catholics. Therefore I think we should (or need) to be as open (or welcoming) to any of them as we could be to any of these potential contributors. I regard this as a priority for now as we don’t have enough contributors (there are 6 member of `romcal` organisation, but currently only you and me work on `romcal`).

> Generally Node.js projects (like that one) on Github don't specify that you have first to install node.js (which include npm). Because people on Github are mostly developpers.
> But I'm aware that `romcal` isn't a "mainstream" project, and might interest people not all developpers or familiar with JavaScript / Node.js development.
_Mostly_ does not mean that there are _only_ devs. Do you regard _reporters_ as devs? I personally not. And although I mostly use Git (any Git server, mostly GitLab) for code versioning, some of my repos contain other, non-code related stuff. Git is for _(text) file versioning_, not only for _code_ versioning.

Therefore I think we should have detailed build instructions. We don’t need to have the build instructions where the regular (`npm`) installation is, but even there should be a link to them. I think we could use GitHub Wiki pages for the documention or have the docs in `docs/` folder within the `romcal` repo. I have made this proposal/suggestion in #57 and #61 is also related to this issue.

> However I dont' think it's a good idea to give all the detailed instructions on how install Node.js, or npm, or Yarn... Because there is various ways in different OS. For exemple I've installed node.js via [`nvm`](https://github.com/nvm-sh/nvm) so I can easily switch between different node.js versions on my mac.
I don’t say we _must_ have a guide how to install `npm` on any and every Linux distribution. A link to a `npm` installation instructions (something like [this](https://nodejs.org/en/download/current/), where the LTS or latest version `npm` can be downloaded, which is recommended way to install, or even using [package managers](https://nodejs.org/en/download/package-manager/). You could even mention the `nvm` which I use too.

> Instead I'll suggest to complete a little bit the "[Usage](https://github.com/romcal/romcal/tree/romcal-es6#usage-)" section in the readme.md, and maybe rename it to Installation instead of Usage (more clear).

I don’t think that the `Usage` is the proper place to include the _build_ instructions—those should be in `Installation` or `Build` instructions. Some projects/programs have an `Install.md` file in the repo root. I think we should have a single place where there we would place all the docs—and I this the best place for them is either `docs/` folder in the repo root or the GitHub Wiki. What do you think?

> You don't have to clone romcal, except if you want to contribute to this project (so theses instructions should be elsewhere in the doc).

You don’t have to clone it unless you want to build the dev version (i.e. one not yet published on `npm`). Some (like me) like to be on the edge for some software to test it and report the issues back. So for testing purposes and when a feature a user wants to use is not yet in the `npm` version yet, the use _has to_ build a program by himself. But the user does not necessarily needs to be a(n `npm`) dev.

> In the contribution section, we should suggest first to fork `romcal` in our own github account, and then clone (to install locally) this forked repo. So you have your own working space (it's a best practice). But I'm pretty sure that this kind of documentation already exists somewhere on Github, so we don't have to reinvent the wheel.

Yes, the _fork → fix → PR → (squash) merge_ should be mentioned in the `Contributing.md` file. Although it might be on GitHub (or elsewhere on the Internet), I’d suggest to place a comprehensive, foolproof guide in `Contributing.md`, because many projects have different requirements. Of course, we need to agree upon one first (we already agreed on the _fork → fix → PR → (squash) merge_ process). Again: the `Contributing.md` does not need to have all the instructions in itself; there can be links to useful sights that have the instructions we need. If there are no such sites (or the instructions are not such as we want), we should place the instructions into the file.

> Also it's no necessary to install `webpack`, `webpack-cli` or `mocha` globally. That should be fixed in this PR.

Good to know. The former instructions (made by @pejulian) included them, so I have kept them.

> And I don't recommend to do `npm audit fix`. Since it could update/modify dependencies of dependencies that could break things. It's best to keep dependencies up to date as far as we can, and if there is critical vulnerabilities, we should report them instead, in the issue tracker of this dependencies.

Agreed on regular patching of the dependencies. The reason I have provided that in the instructions is that I have always had problems with `romcal` installation (although I have never tried to do so on vanilla OS). Other users might get the same errors, so I think there should be at least a remark how to solve this _potential_ error.

> ### 2
> I don't have this issue one Firefox, Safari or Chrome on my mac. But there is still some fine tuning on the build steps for the web package of `Romcal`. Probably in another PR.

I use Firefox Nightly (current version 72.0a1) on CentOS 7 x86_64. But it looks like `romcal` is buildable, so we might postpone it to a later PR.

> ### 3
> Here it's just to give a very basic exemple, that we can easily hack directly in the browsers. So I don't think it's a problem if it don't follow good practices in web development here.

I don’t say it is a _problem_, but I think it might be regarded as _enhancement_. We might push the improvement of the example to a later PR. (I think the example might show all/most features `romcal` offers.)

> ### 4
> I already created a new version `1.3.1`. The major refactor from @pejulian correspond to `1.3.0` (also not published yet on NPM).

I think that by _releasing_ we talk about different things. I think a _proper_ release (and all release in general, be it major, minor or patch) should have the following (let’s say) attributes after they are merge to `master`:
- a tag (like [`v1.2.0`](https://github.com/romcal/romcal/tree/v1.2.0) has);
- a GitHub release (like [`v1.2.0`](https://github.com/romcal/romcal/releases/tag/v1.2.0) has);
- an `npm` release (like [`v1.2.3`](https://www.npmjs.com/package/romcal) has).

I regard this as proper versioning system. And I think we should never skip a version publishing to `npm`, therefore only release (i.e. merge to `master`) a new `romcal` version when everything has be double-checked. :smiley:

I agree that this is only a patch to `v1.3.0`, so no need to increment the major version number. And it is also true that @pejulian did not had time to make proper releases at the end of his maintainership (before he went AWOL).

And I would like to note that we should include proper _release notes_ what has changed since the last release. Currently, you merge directly to `master` (sort of) any single commit. I think it would be better to have a dev branch were we would check and test if everything is working and then do a squash merge of all commits with release notes in the merge description. The `git log` would be usable and pretty again. :smiley:

> To stay in the scope of this PR, the following points you are mentioning (as well as the installation documentation) should be handled in separated PRs.

Agreed.

> Travis: I have not looked at it yet. I know it was working and checking automatically all new PR for this project, a few months (years) ago. But I don't know what's going on today. Yes we should took a look into it as well...

I have no idea too. But it might related to my previous findings (#66 and #67) that the OS used at Travis is outdated. But we might need some authorisation from @pejulian, as if I recall correctly, I could not change the config (because I don’t have @pejulian’s credentials). But I have no experience with Travis. :smiley:

---

Today I have found another issue (I regard it as major bug): when I change the locale in the example to `sk` (Slovak), most strings are not localised (although they _are_ localised in `skSk.js`). For example:

```bash
300: Object { type: "FEAST", name: "Saints Simon and Jude, Apostles", key: "saintsSimonAndJudeApostles", … }
301: Object { type: "WEEKDAY", name: "utorok of the 30. week of Ordinary Time", key: "utorokOfThe30WeekOfOrdinaryTime", … }
```

emagnier commented:
I agreed with you 👍 

For the release, I still thing working directly on `master` is enough for small things. People will don't get automatically the update until we tag and publish to npm, and we can always take care of the release # and history before publish it. This is the way a lot other projects are doing, and in my experience point of view, I thing it's also reasonable for `romcal`. 
If people include romcal in their projects, targeting the last commit change instead, instead a last/specific release (which is a bad practice unless you know what they are doing), it's at their own risk.

But let's try to keep the discussion on the scope of this PR, and use the issue tracker for all other things.

tukusejssirs commented:
> For the release, I still thing working directly on `master` is enough for small things.

Okay, but we should not increment a version when we don’t publish one. But this seems to me not the proper way to do releases.

tukusejssirs commented:
Still, I don’t know if you have read this (I think this is very related to this PR):

> Today I have found another issue (I regard it as major bug): when I change the locale in the example to sk (Slovak), most strings are not localised (although they are localised in skSk.js). For example:
>
> ```bash
> 300: Object { type: "FEAST", name: "Saints Simon and Jude, Apostles", key: "saintsSimonAndJudeApostles", … }
> 301: Object { type: "WEEKDAY", name: "utorok of the 30. week of Ordinary Time", key: "utorokOfThe30WeekOfOrdinaryTime", … }
> ```

emagnier commented:
We are not yet on master, so not the time yet to do a release. And we don't bump the version each time we do a commit. The purpose of this PR was to make the `romcal-es6` branch working again so then we can test and fix what need to be fixed.

For the bug you mentionned, please create a new issue (if it doesn't exist) so we can check that separately. I'm actually rebasing this branch from `master` and I'm getting some headaches with locale keys since we got updates in both `master` and `romcal-es6`. It have too much stuffs not related of the main purpose of this branch: revamp the code to the es6 syntax with the help babel, esm and webpack.

tukusejssirs commented:
I believe that the version bumping should be done _after_ all changes are made and were checked, not _before_. That is, the version bumping should be the last thing made before version release (merging to `master`, GitHub tag and release creation and `npm` release).

I see that too many things needs to be fixed at once. Still I am glad you’ve taken up that burden. ;)

I have already created/updated some issues based upon our conversation in this PR. :smiley:
